### PR TITLE
Fix @count on polymorphic relations

### PR DESCRIPTION
**Pull request by tlaverdure**
- [X] Added or updated tests
- [ ] Documented user facing changes
- [X] Updated CHANGELOG.md

When using `@count` on a model that morphsTo different models, the batch loader may confuse keys between the different models.

Related: https://github.com/laravel/framework/pull/25997#issue-221049454

**Changes**

Separate models that need to be counted by their class name before attempting to load counts.

**Breaking changes**

N/A

### CODE REVIEW

**Comment by spawnia** [comment]
Will review later, but for now let me just shamelessly piggyback on the fact that you can deal with `@count` and the likes of it will and ask you to take a look at https://github.com/nuwave/lighthouse/issues/1461 😅

**Comment by tlaverdure** [comment]
Cool, thanks. I'll see if I can help.

**Comment by tlaverdure** [comment]
@spawnia got some time to take a look at this? Thanks

File: src/Execution/DataLoader/RelationCountBatchLoader.php [hunk 1]
```diff
@@ -39,23 +40,27 @@ public function __construct(
...
-    protected function getParentModels(): EloquentCollection
+    protected function getParentModels(): Collection
     {
-        return new EloquentCollection(
-            array_map(
-                function (array $meta) {
+        return Collection::make($this->keys)
+            ->groupBy(function ($key) {
+                return get_class($key['parent']);
+            }, true)
```
Comment (spawnia) [inline src/Execution/DataLoader/RelationCountBatchLoader.php:56]:
```suggestion
            ->groupBy(
                static function (array $key): string {
                    return get_class($key['parent']);
                },
                true
            )
```

File: src/Execution/DataLoader/RelationCountBatchLoader.php [hunk 2]
```diff
@@ -39,23 +40,27 @@ public function __construct(
...
+            ->mapWithKeys(function ($keys) {
```
Comment (spawnia) [inline src/Execution/DataLoader/RelationCountBatchLoader.php:57]: Strict type hints everywhere.

File: src/Execution/DataLoader/RelationCountBatchLoader.php [hunk 3]
```diff
@@ -39,23 +40,27 @@ public function __construct(
...
     }
 
     /**
-     * Get the parents from the keys that are present on the BatchLoader.
+     * Get the parent models from the keys that are present on the BatchLoader.
+     * Models are grouped by their fully qualified class name to prevent key
+     * collisions between different types of models.
      */
-    protected function getParentModels(): EloquentCollection
+    protected function getParentModels(): Collection
```
Comment (spawnia) [inline src/Execution/DataLoader/RelationCountBatchLoader.php:51]: What's in the collection? Please add a type hint such as `@return Collection<?>`

File: src/Execution/Utils/ModelKey.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * Create a model key that concatenates the models fully-qualified class
+ * name and key or composite key.
+ *
+ * For example: App\Model\Post:1
+ */
+class ModelKey
+{
+    public static function build(Model $model): string
+    {
+        return Collection::make([get_class($model)])
```
Comment (spawnia) [inline src/Execution/Utils/ModelKey.php:18]:
```suggestion
        return (new Collection([get_class($model)]))
```

Comment (spawnia) [reply src/Execution/Utils/ModelKey.php:18]: Actually, can you write that function without collections? This code instantiates three Collection objects to do something quite simple. Performance matters.

File: tests/Unit/Schema/Directives/CountDirectiveTest.php [hunk 5]
```diff
@@ -174,4 +194,73 @@ public function testItCountsARelationshipWithScopesApplied(): void
...
+            }
+        }
+        ')->assertExactJson([
+            'data' => [
+                'activity' => [
+                    [
+                        'id' => '1',
+                        'content' => [
+                            'id' => "{$post1->id}",
+                            'images_count' => $post1->images->count(),
```
Comment (spawnia) [inline tests/Unit/Schema/Directives/CountDirectiveTest.php:245]:
```suggestion
                            'images_count' => $post1->images()->count(),
```

File: tests/Utils/Models/Activity.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Relations\MorphTo;
+
+/**
+ * @property int $id
+ * @property int $activity_id
+ * @property string $activity_type
+ * @property \Carbon\Carbon $created_at
+ * @property \Carbon\Carbon $updated_at
```
Comment (spawnia) [inline tests/Utils/Models/Activity.php:14]:
```suggestion
 * @property \Carbon\Carbon $updated_at
 *
 * Relations
 * @property-read \Tests\Utils\Models\Post|\Tests\Utils\Models\Task $content
 * @property-read \Tests\Utils\Models\User $user
```

Also, please add `|null` where appropriate.

File: tests/Utils/Models/Post.php [hunk 7]
```diff
@@ -30,6 +30,11 @@ public function user(): BelongsTo
         return $this->belongsTo(User::class);
     }
 
+    public function activity(): MorphMany
```
Comment (spawnia) [inline tests/Utils/Models/Post.php:33]: This needs a `@property-read`.

File: tests/database/migrations/2020_07_09_000000_create_testbench_activities_table.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class CreateTestbenchActivitiesTable extends Migration
+{
+    /**
+     * Run the migrations.
+     */
```
Comment (spawnia) [inline tests/database/migrations/2020_07_09_000000_create_testbench_activities_table.php:11]:
```suggestion
```

File: tests/Utils/Models/Activity.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ */
+class Activity extends Model
+{
+    /**
+     * The attributes that aren't mass assignable.
+     *
+     * @var array<string>
+     */
+    protected $guarded = [];
+
```
Comment (spawnia) [inline tests/Utils/Models/Activity.php:24]:
```suggestion
```

Lighthouse no longer respects mass-assignment protection 😉

File: src/Execution/Utils/ModelKey.php (hunk 4 above)
Comment (tlaverdure) [reply src/Execution/Utils/ModelKey.php:18]: You got it.

File: tests/Utils/Models/Post.php (hunk 7 above)
Comment (spawnia) [reply tests/Utils/Models/Post.php:33]: What i meant is that the PHPDoc of the model is missing an annotation that makes sure that the magic relation access is typesafe.

File: tests/Utils/Models/Task.php [hunk 10]
```diff
@@ -37,6 +37,11 @@ protected static function boot(): void
         });
     }
 
+    public function activity(): MorphMany
```
Comment (spawnia) [inline tests/Utils/Models/Task.php:40]: This model also lacks a PHPDoc annotation for this relation.

File: tests/Utils/Models/User.php [hunk 11]
```diff
@@ -23,6 +23,7 @@
  *
  * @property-read \Illuminate\Database\Eloquent\Collection<\Tests\Utils\Models\Task> $tasks
  * @property-read \Illuminate\Database\Eloquent\Collection<\Tests\Utils\Models\Post> $posts
+ * @property-read \Tests\Utils\Models\Activity $activity
```
Comment (spawnia) [inline tests/Utils/Models/User.php:26]: This does not seem right.

Comment (tlaverdure) [reply tests/Utils/Models/User.php:26]: oops that should be on `Post`

File: tests/Utils/Models/Task.php (hunk 10 above)
Comment (tlaverdure) [reply tests/Utils/Models/Task.php:41]: Sorry, added.

**Pull request merged**